app/celery/process_pinpoint_receipts_tasks.py

```diff
@@ -0,0 +1,150 @@
+            return
+
+        if notification.status != NOTIFICATION_SENT:
+            notifications_dao._duplicate_update_warning(notification, notification_status)
+            return
+
+        notifications_dao._update_notification_status(
+            notification=notification,
+            status=notification_status,
+            provider_response=provider_response,
+        )
+
+        if notification_status != NOTIFICATION_DELIVERED:
+            current_app.logger.info(
+                (
+                    f"Pinpoint delivery failed: notification id {notification.id} and reference {reference} has error found. "
+                    f"Provider response: {provider_response}"
+                )
+            )
+        else:
+            current_app.logger.info(
+                f"Pinpoint callback return status of {notification_status} for notification: {notification.id}"
+            )
+
+        statsd_client.incr(f"callback.sns.{notification_status}")  # TODO: do we want a Pinpoint metric here?
+
+        if notification.sent_at:
+            statsd_client.timing_with_dates(
+                "callback.sns.elapsed-time", datetime.utcnow(), notification.sent_at
+            )  # TODO: do we want a Pinpoint metric here?
+
+        _check_and_queue_callback_task(notification)
+
+        return True
+
+    except Retry:
+        raise
+
+    except Exception as e:
+        current_app.logger.exception(f"Error processing Pinpoint results: {str(e)}")
+        self.retry(queue=QueueNames.RETRY)
+
+
+def determine_pinpoint_status(status: str, provider_response: str) -> Union[str, None]:
+    """Determine the notification status based on the SMS status and provider response.
+
+    Args:
+        status (str): message status from AWS
+        provider_response (str): detailed status from the SMS provider
+
+    Returns:
+        Union[str, None]: the notification status or None if the status is not handled
+    """
+
+    if status == "DELIVERED":
+        return NOTIFICATION_DELIVERED
+
+    response_lower = provider_response.lower()
+    match response_lower:
+        case response_lower if "blocked" in response_lower:
+            return NOTIFICATION_TECHNICAL_FAILURE
+        case response_lower if "invalid" in response_lower:
+            return NOTIFICATION_TECHNICAL_FAILURE
+        case response_lower if "is opted out" in response_lower:
+            return NOTIFICATION_PERMANENT_FAILURE
+        case response_lower if "unknown error" in response_lower:
+            return NOTIFICATION_TECHNICAL_FAILURE
+        case response_lower if "exceed max price" in response_lower:
+            return NOTIFICATION_TECHNICAL_FAILURE
+        case "phone carrier is currently unreachable/unavailable":
+            return NOTIFICATION_TEMPORARY_FAILURE
+        case "phone is currently unreachable/unavailable":
+            return NOTIFICATION_PERMANENT_FAILURE
+
+    return None
```

Comment:
I think this is the cleanest way to use matching, but I think it's even better with if/elifs:
```
    if "blocked" in response_lower:
        return NOTIFICATION_TECHNICAL_FAILURE
    elif "invalid" in response_lower:
        return NOTIFICATION_TECHNICAL_FAILURE
    elif "is opted out" in response_lower:
        return NOTIFICATION_PERMANENT_FAILURE
    elif "unknown error" in response_lower:
        return NOTIFICATION_TECHNICAL_FAILURE
    elif "exceed max price" in response_lower:
        return NOTIFICATION_TECHNICAL_FAILURE
    elif "phone carrier is currently unreachable/unavailable" in response_lower:
        return NOTIFICATION_TEMPORARY_FAILURE
    elif "phone is currently unreachable/unavailable" in response_lower:
        return NOTIFICATION_PERMANENT_FAILURE
    else:
        return None
```

Reply:
Yeah the match case isn't the most terse I have seen in a language unfortunately. I hope Python improves on that. I wouldn't mind the `if...elif...else`. :\